### PR TITLE
table: add `WritableConstraint` to interface `table.Table` to avoid implicit interface implementation

### LINKED ISSUE #54824: Add `WritableConstraint` to `table.Table` to avoid implicit interface implementation
## Enhancement

See description in PR: #54825

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #54824

Problem Summary:

An implicit interface `CheckConstraintTable` should be implemented for all table objects to check constraints somewhere. For example: 

https://github.com/pingcap/tidb/blob/f774ef6fb5bd140c4db67a1ab5fc3dea53253bec/pkg/table/tables/partition.go#L1582-L1593

However, it is not a good way to achieve it. The coders unfamiliar with those codes may rename or remove the method 'CheckRowConstraint` by mistake, but it will not produce any error in the build phase.

To make the code strong, it's better not to use implicit interface implementation and all methods a table should implement should be declared in `table.Table`

### What changed and how does it work?

1. Add `WritableConstraint` to `table.Table` and its derived type should implement it.
2. Add static functions `CheckRowConstraint` and `CheckRowConstraintWithDatum` to check constraints.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by YangKeao — APPROVED** [review]
LGTM

File: pkg/table/table.go [hunk 1]
```diff
@@ -303,3 +308,26 @@ type CachedTable interface {
...
+	}
+	return nil
+}
+
+// CheckRowConstraintWithDatum verify row check constraints.
+// It is the same with `CheckRowConstraint` but receives a slice of `types.Datum` instead of `chunk.Row`.
+func CheckRowConstraintWithDatum(ctx exprctx.EvalContext, constraints []*Constraint, row []types.Datum) error {
+	if len(constraints) == 0 {
+		return nil
+	}
```
Comment (Defined2014) [inline pkg/table/table.go:331]: Do we need this? `CheckRowConstraint` also return nil when `len(constraints) == 0`.

**Review by Defined2014 — APPROVED** [review]
reset LGTM

File: pkg/table/table.go (hunk 1 above)
Comment (lcwangchao) [reply pkg/table/table.go:331]: Yes, to avoid unnecessary perf cost of `chunk.MutRowFromDatums(row).ToRow()`

**Review by tangenta — APPROVED** [review]

**Pull request merged**